src/MongoDB.Analyzer/Core/Builders/BuildersResolveVariables.cs

```diff
@@ -0,0 +1,445 @@
+
+            if (!canEvaluate)
+            {
+                var leftEvaluationResult = PreEvaluate(binaryExpressionSyntax.Left, contextInformation, syntaxNodeToVariableValueMap : null, displayDiagnostics);
+
+                if (leftEvaluationResult.CanEvaluate)
+                {
+                    var rightEvaluationResult = PreEvaluate(binaryExpressionSyntax.Right, contextInformation, syntaxNodeToVariableValueMap : leftEvaluationResult.SyntaxNodeToVariableValueMap, displayDiagnostics);
+                    canEvaluate = leftEvaluationResult.CanEvaluate && rightEvaluationResult.CanEvaluate;
+                    syntaxNodeToVariableValueMap = rightEvaluationResult.SyntaxNodeToVariableValueMap;
+                }
+            }
+            else
+            {
+                syntaxNodeToVariableValueMap ??= new Dictionary<SyntaxNode, VariableValue>();
+                var rightVariableValue = new VariableValue(expressionAnalysisContext, expressionAnalysisContext.Node.Locations.ToList());
+                syntaxNodeToVariableValueMap.Add(binaryExpressionSyntax, rightVariableValue);
+            }
+        }
+
+        if (isCompoundExpression && canEvaluate)
+        {
+            var assignmentExpression = syntaxNode as AssignmentExpressionSyntax;
+            var leftHandSideIdentifier = assignmentExpression.Left.TrimParenthesis() as IdentifierNameSyntax;
+            canEvaluate &= variableValues.TryGetValueSafe(leftHandSideIdentifier.Identifier.ValueText, out var leftVariableValue);
+
+            if (canEvaluate)
+            {
+                syntaxNodeToVariableValueMap ??= new Dictionary<SyntaxNode, VariableValue>();
+                syntaxNodeToVariableValueMap.Add(leftHandSideIdentifier, leftVariableValue);
+            }
+        }
+
+        return new EvaluationResult(canEvaluate, syntaxNodeToVariableValueMap);
+    }
+
+    private static ExpressionAnalysisContext GenerateNewExpressionAnalysisContext(
+        SyntaxNode syntaxNode,
+        Dictionary<SyntaxNode, VariableValue> builderNodeToVariableValue)
+    {
+        var syntaxKind = syntaxNode.Kind();
+        var argumentTypeName = "";
+        var constantMappers = new List<ConstantsMapper>();
+        var nodesRemapping = new Dictionary<SyntaxNode, SyntaxNode>();
+
+        foreach (var builderNode in builderNodeToVariableValue.Keys)
+        {
+            var variableValue = builderNodeToVariableValue[builderNode];
+            var analysisContext = variableValue.ExpressionAnalysisContext;
+            var rewrittenBuildersNode = analysisContext.Node.RewrittenExpression;
+            var constantsMapper = analysisContext.Node.ConstantsRemapper;
+            argumentTypeName = analysisContext.Node.ArgumentTypeName;
+            nodesRemapping.Add(builderNode, rewrittenBuildersNode);
+            constantMappers.Add(constantsMapper);
+        }
+
+        var rewrittenSyntaxNode = syntaxNode.ReplaceNodes(nodesRemapping.Keys, (n, _) => nodesRemapping[n]);
```

Comment:
We can skip the new dictionary instance
`syntaxNode.ReplaceNodes(builderNodeToVariableValue.Keys, (n, _) => builderNodeToVariableValue[n].ExpressionAnalysisContext.Node.RewrittenExpression)`

Reply:
Done